src/JWT.ts

```diff
@@ -392,61 +493,99 @@ export async function verifyJWT(
   const proofPurpose: ProofPurposeTypes | undefined = Object.prototype.hasOwnProperty.call(options, 'auth')
     ? options.auth
       ? 'authentication'
       : undefined
     : options.proofPurpose
 
-  let did
+  let didUrl: string | undefined
 
   if (!payload.iss && !payload.client_id) {
     throw new Error(`${JWT_ERROR.INVALID_JWT}: JWT iss or client_id are required`)
   }
 
-  if (payload.iss === SELF_ISSUED_V2 || payload.iss === SELF_ISSUED_V2_VC_INTEROP) {
+  if (options.didAuthenticator) {
+    didUrl = options.didAuthenticator.issuer
+  } else if (payload.iss === SELF_ISSUED_V2 || payload.iss === SELF_ISSUED_V2_VC_INTEROP) {
     if (!payload.sub) {
       throw new Error(`${JWT_ERROR.INVALID_JWT}: JWT sub is required`)
     }
     if (typeof payload.sub_jwk === 'undefined') {
-      did = payload.sub
+      didUrl = payload.sub
     } else {
-      did = (header.kid || '').split('#')[0]
+      didUrl = (header.kid || '').split('#')[0]
     }
   } else if (payload.iss === SELF_ISSUED_V0_1) {
     if (!payload.did) {
       throw new Error(`${JWT_ERROR.INVALID_JWT}: JWT did is required`)
     }
-    did = payload.did
+    didUrl = payload.did
   } else if (!payload.iss && payload.scope === 'openid' && payload.redirect_uri) {
     // SIOP Request payload
     // https://identity.foundation/jwt-vc-presentation-profile/#self-issued-op-request-object
     if (!payload.client_id) {
       throw new Error(`${JWT_ERROR.INVALID_JWT}: JWT client_id is required`)
     }
-    did = payload.client_id
+    didUrl = payload.client_id
   } else {
-    did = payload.iss
+    didUrl = payload.iss
   }
 
-  if (!did) {
+  if (!didUrl) {
     throw new Error(`${JWT_ERROR.INVALID_JWT}: No DID has been found in the JWT`)
   }
 
-  const { didResolutionResult, authenticators, issuer }: DIDAuthenticator = await resolveAuthenticator(
-    options.resolver,
-    header.alg,
-    did,
-    proofPurpose
-  )
-  const signer: VerificationMethod = await verifyJWSDecoded({ header, data, signature } as JWSDecoded, authenticators)
-  const now: number = typeof options.policies?.now === 'number' ? options.policies.now : Math.floor(Date.now() / 1000)
-  const skewTime = typeof options.skewTime !== 'undefined' && options.skewTime >= 0 ? options.skewTime : NBF_SKEW
+  let authenticators: VerificationMethod[]
+  let issuer: string
+  let didResolutionResult: DIDResolutionResult
+  if (options.didAuthenticator) {
+    ;({ didResolutionResult, authenticators, issuer } = options.didAuthenticator)
+  } else {
+    ;({ didResolutionResult, authenticators, issuer } = await resolveAuthenticator(
+      options.resolver,
+      header.alg,
+      didUrl,
+      proofPurpose
+    ))
+    // Add to options object for recursive reference
+    options.didAuthenticator = { didResolutionResult, authenticators, issuer }
+  }
+
+  const { did } = parse(didUrl) as ParsedDID
+
+  let signer: VerificationMethod | null = null
+
+  if (did !== didUrl) {
```

Comment:
This comparison assumes that a didUrl that does not match the parsed DID refers to one of the verification methods.
The assumption will break if the didUrl contains, for example, query parameters.

Example: `did:ethr:0xabcd...?versionId=1234`.
This is a valid issuer with a DID document potentially different from `did:ethr:0xabcd...`

Is the intent here to avoid looping through the `authenticators` array?

Reply:
if the issuer of the DID is a DID URL, here we assume that it is pointing to a specific verification method in the DID Document. This is used in the recursive calling to `verifyJWT()` for delegation

`did:ethr` is not a good example, as it cannot express multi-sig conditions or delegations.

Take for example two DIDs which one verification method on each:
1. `did:antelope:jack#delegated` which delegates to `did:antelope:thomas#base`
2. `did:antelope:thomas#base` with a ES256K keypair

A VC is issued by `did:antelope:jack#delegated`. The VC is signed by the  `did:antelope:thomas#base` keypair.

- `verifyJWT()` is called which finds the issuer to be `did:antelope:jack#delegated`
- it goes and finds the specific verification method, which tells that it is delegating to `did:antelope:thomas#base`
- `verifyJWT()` is then called recursively to check whether the signature matches the `did:antelope:thomas#base` which it does.

It is important to note that at each of the above steps, a specific verification method is found using the DID URL, required to make the above flow work.

Reply:
Thank you for clarifying the intent.
I realize did:ethr is not relevant to the conditional verification, but the changes I'm asking about are relevant for all verification, not just conditional.

I'll add a test for the case I'm referring to.

Reply:
Yes I see your consern. If someone issues using a DIDUrl that does not point directly to a verification method, it will now fail. Correct?

Reply:
I'm thinking of ways around this, perhaps we can change it so that it only checks the fragment part of the DID URL against the fragment of the verification method. What do you think?

Reply:
That's one possibility.
Another option would be to just try to find the verification method, which would work for the conditional logic you need, but then to not fail if it's not found, which would continue to do the verification as before.

Reply:
JWT.ts verifyJWT()

```ts
  const { fragment } = parse(didUrl) as ParsedDID

  let signer: VerificationMethod | null = null

  // If a fragment is supplied, assume it is pointing to a specific Verification Method
  // This is a required feature for Conditional Proof delegated signatures
  if (fragment) {
    const authenticator = authenticators.find((auth) => (parse(auth.id) as ParsedDID).fragment === fragment)
 ```

this would avoid this situation (not 100%, e.g. if someone used fragments to refer to a non verification method it would still break)

Reply:
Your solution with the fragment seems like the right approach, but somehow comparing just the fragment parts seems odd, as it seems close to being able to impersonate some other DID.

How about this pseudocode?
```
if (didURL contains fragment) {
  find authenticator that matches didUrl
} else {
  try to check the proof against all authenticators
}
```

Reply:
the original intent, was to treat the DID URL as a relative URL, which are assumed to dereference to a resource in the DID Document

Reply:
OK, that's even better, as it would also cover the case where the issuer is a URL but there's also a `kid` specified in the header, relative to the issuer URL.

I'll merge this PR as is and make the relative `kid` changes as a separate item.